### DISCUSSION TITLE
retornar index de array fora do map

### DISCUSSION THREAD

**Discussion by thauealfredo** [frontendbr/forum#1938] [Perguntas e Respostas]
Galera eu tenho um componente que monta linhas da tabela de acordo com um array que eu passo. Dentro desse array eu passo uma img com onClick, eu queria saber como posso pegar o index equivalente clicado

o item do array
![image](https://user-images.githubusercontent.com/50925608/101826749-32a22480-3b0e-11eb-9018-8943c91eecb8.png)

Esse é o componente 
![image](https://user-images.githubusercontent.com/50925608/101826668-143c2900-3b0e-11eb-8d11-c1b4c88646c0.png)

**Comment by ninetails** [comment]
[O segundo argumento do _callback_ do Array.prototype.map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#Parameters) é o índice do array. Vc pode repassar ele até a chamada da função no `onClick`

```jsx
{arr.map((item, index) => (
  <Componente onClick={event => meuHandler(event, index)}>
    ...
  </Componente>
))}

```

**Reply by thauealfredo** [reply]
Se eu por dessa forma em todas as colunas vai ter onClick e n é isso que eu quero

**Reply by ninetails** [reply]
<details>
<summary>Vê se é isso</summary>

```jsx
const customHandler = (event, row, col) =>
  console.log(`Handler ${row}-${col}`, event);

const arr = [
  [
    {
      id: "action",
      value: (row, col) => (
        <button onClick={(event) => customHandler(event, row, col)}>
          <img
            src="https://via.placeholder.com/150?text=1-1"
            alt="Imagem 1-1"
          />
        </button>
      )
    },
    {
      id: "action",
      value: (row, col) => (
        <button onClick={(event) => customHandler(event, row, col)}>
          <img
            src="https://via.placeholder.com/150?text=1-2"
            alt="Imagem 1-2"
          />
        </button>
      )
    }
  ],
  [
    {
      id: "action",
      value: (row, col) => (
        <button onClick={(event) => customHandler(event, row, col)}>
          <img
            src="https://via.placeholder.com/150?text=2-1"
            alt="Imagem 2-1"
          />
        </button>
      )
    },
    {
      id: "action",
      value: (row, col) => (
        <button onClick={(event) => customHandler(event, row, col)}>
          <img
            src="https://via.placeholder.com/150?text=2-2"
            alt="Imagem 2-2"
          />
        </button>
      )
    }
  ]
];

export default function App() {
  return (
    <>
      <table>
        <tbody>
          {arr &&
            arr.map((row, rowNum) => (
              <tr key={rowNum}>
                {row.map((col, colNum) => (
                  <td key={colNum}>{col.value(rowNum, colNum)}</td>
                ))}
              </tr>
            ))}
        </tbody>
      </table>
    </>
  );
}

```
</details>

[![Edit late-leftpad-wjsb2](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/late-leftpad-wjsb2?fontsize=14&hidenavigation=1&theme=dark)

**Reply by thauealfredo** [reply]
Eu achei um jeito, com uma gambiarra, testei com seu código mas ai n renderizou o botão. mas  valeu

**Reply by wilsonneto-swe** [reply]
Xiii, gambiarra não heim rs Mas o correto pelo que você @thauealfredo  descreveu, é como o @ninetails  falou mesmo, você usa o segundo parametro do map... E só irá ter em todas as colunas se você colocar em todas as colunas. E o melhor dependendo de sua tabela é usar o id da entidade que está naquela linha e não o indice, dependendo do cenário.
Abraços!

**Reply by thauealfredo** [reply]
Como não dava para passar o index da forma que eu estava fazendo, eu fiz uma função que passo por callback, e nessa função eu retornava o elemento clicado, o pai desse elemento e o id dele que era o index(recurso técnico alternativo xD), mas só agora, no meio do banho, me dei conta que eu consigo passar o index nessa função de callback

**Reply by wilsonneto-swe** [reply]
Show, penso que era isto mesmo que o @ninetails mostrou ali, que bom que rolou ai, abraços!

**Reply by ninetails** [reply]
assim, ideal ideal mesmo seria vc botar um ID único na lista, pois mesmo o React não recomenda utilizar `key` em loops.

```js
const arr = [
  // ...
  {
    id: "ID ÚNICO, TALVEZ UM HASH",
    value: (row, col) => (
      <button onClick={(event) => customHandler(event, row, col)}>
        <img
          src="https://via.placeholder.com/150?text=1-1"
          alt="Imagem 1-1"
        />
      </button>
    )
  },
  // ...
]
```

E aí ao remover vc manipula o array passando esse ID.
